Source/ARTLocalDevice.m

```diff
@@ -102,6 +95,33 @@ + (ARTLocalDevice *)load:(NSString *)clientId storage:(id<ARTDeviceStorage>)stor
     return device;
 }
 
+- (BOOL)setupDetailsWithClientId:(NSString *)clientId {
+    NSString *deviceId = self.id;
+    NSString *deviceSecret = self.secret;
+    
+    if ([deviceId isEqualToString:@""] || deviceSecret == nil) { // generate both at the same time
+        deviceId = [self.class generateId];
+        deviceSecret = [self.class generateSecret];
+        
+        [_storage setObject:deviceId forKey:ARTDeviceIdKey];
+        [_storage setSecret:deviceSecret forDevice:deviceId];
+    }
+    
+    self.id = deviceId;
+    self.secret = deviceSecret;
+    
+    NSString *localClientId = self.clientId;
+    
+    if (localClientId && clientId && ![localClientId isEqualToString:clientId]) {
+        ARTLogError(self.logger, @"ARTLocalDevice: `clientId`s don't match: %@ vs %@.", clientId, localClientId);
```

Comment:
What spec point does this correspond to?

Reply:
This is for [RSH3a2a1](https://sdk.ably.com/builds/ably/specification/main/features/#RSH3a2a1). I've avoided to make further changes in this PR due to some urgency for the customer, but in case of mismatched clientId I would generate error [earlier](https://github.com/ably/ably-cocoa/blob/6c6f0fbd12da95c1b47bc474f791fbd854ec3ff0/Source/ARTPushActivationState.m#L121) then it's implemented now in `validateAndSync` function. That's why the return value is needed.

Reply:
But I don't understand what this new `clientId` check that you've introduced in the current PR is achieving?

Reply:
agree, better to add all of this when needed - https://github.com/ably/ably-cocoa/pull/1847/commits/c64c56192644dfc2a865c2b41a411be72d2bfb71